service/cachemdw/middleware.go

```diff
@@ -0,0 +1,109 @@
+package cachemdw
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+
+	"github.com/kava-labs/kava-proxy-service/decode"
+)
+
+type cacheContextKey string
+
+const (
+	CachedContextKey cacheContextKey = "X-KAVA-PROXY-CACHED"
+
+	CacheHeaderKey       = "X-Cache"
+	CacheHitHeaderValue  = "HIT"
+	CacheMissHeaderValue = "MISS"
+)
+
+func (c *ServiceCache) Middleware(
+	next http.Handler,
+) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		uncachedContext := context.WithValue(r.Context(), CachedContextKey, false)
+
+		req := r.Context().Value(c.decodedRequestContextKey)
+		decodedReq, ok := (req).(*decode.EVMRPCRequestEnvelope)
+		if !ok {
+			next.ServeHTTP(w, r.WithContext(uncachedContext))
+			return
+		}
+
+		cachedQueryResponse, err := c.GetCachedQueryResponse(r.Context(), decodedReq, r.Host)
+		if err == nil {
```

Comment:
normal control flow shouldn't happen inside the if, it should be if err !=nil we handle the exceptional case, otherwise the code reads top to bottom for the nominal code path.

Reply:
```go
cachedQueryResponse, err := c.GetCachedQueryResponse(r.Context(), decodedReq, r.Host)
if err == nil {
	c.respond(next, w, r, cachedQueryResponse)
	return
}

// a lot of business logic...
```

not fully sure how to refactor this, I can do smth like this:
```go
cachedQueryResponse, err := c.GetCachedQueryResponse(r.Context(), decodedReq, r.Host)
if err != nil && err != ErrNotFound {
   logError()
   return
}
if err == ErrNotFound {
	c.respond(next, w, r, cachedQueryResponse)
	return
}

// a lot of business logic...
```

Reply:
good refactor idea

Reply:
it was automatically refactored after I split up middleware into 2 parts, now it looks like this:
```go
cachedQueryResponse, err := c.GetCachedQueryResponse(r.Context(), decodedReq)
if err != nil {
	next.ServeHTTP(w, r.WithContext(uncachedContext))
	return
}

// a lot of business logic...
```